src/cata_tiles.h

```diff
@@ -25,9 +25,37 @@ class JsonObject;
 struct visibility_variables;
 
 /** Structures */
+struct sprite_variation {
+    int weight;
+    std::vector<int> sprite_ids;
+
+    sprite_variation()
+    {
+        weight = 1;
+        sprite_ids.clear();
+    }
+
+    sprite_variation(int id)
+    {
+        weight = 1;
+        sprite_ids.clear();
+        sprite_ids.push_back(id);
+    }
+
+    sprite_variation(int w, int id)
+    {
+        weight = w;
+        sprite_ids.clear();
+        sprite_ids.push_back(id);
+    }
+};
+
 struct tile_type
 {
-    std::vector<int> fg, bg;
+    // first entry of each of these vectors is not a real sprite_variation
+    // but a bookkeeping list instead, used for reducing the task of selecting
+    // a random variation from O(N) to O(1)
```

Comment:
Put `fg` and bookkeeping thing into another class. You could even make this new class derive from `std::vector<int>`, so existing code that expects and reads a `std::vector<int>` will continue to work.

Actually this could be quite useful as `fg` and `bg` seem to have the same structure, the new class could be used for both.